Fix for when a list element is None from do_meta

juhaa opened the pull request:
Fixes #23

juhaa commented:
Or maybe better to change the output from `inv_var_meta` and `variance_weight_meta` to be consistent?
Instead of
```
return (beta_meta, math.sqrt(1/sum_inv_var), max(sys.float_info.min * sys.float_info.epsilon, 2 * scipy.stats.norm.sf(abs(sum(effs_inv_var) / math.sqrt(sum_inv_var) ))), het_p) if len(effs_inv_var)==len(studies) else None
```
just
```
return (beta_meta, math.sqrt(1/sum_inv_var), max(sys.float_info.min * sys.float_info.epsilon, 2 * scipy.stats.norm.sf(abs(sum(effs_inv_var) / math.sqrt(sum_inv_var) ))), het_p)
```

juhis reviewed (approved):
Nice!

Fedja commented:
how did we never have this before. I remember adding that check there when there were 0 SE in some studies... maybe evolved in later pull. Anyway good to add now. While you are at it can you add doc-string to those functions documenting the behaviour?

juhaa commented:
There is a check for 0 or missing SE but the problem was the improper handling of the outputs from meta-analysis method functions. I think it's better convention for the functions to return a fixed output rather than e.g. a tuple or just None, as it was in this case. Also, the heterogeneity test result was not handled properly when generating the outputs so I added checks for this.

So, I added couple commits where I:
1. Fixed the output from meta-analysis method functions to tuple (tuple of length 3 or 4 whether het test was calculated)
2. Added checks if het test was performed and adjust output formatting accordingly
3. Added docstrings and output type hints for the method functions

juhis reviewed (approved):
Great!

Fedja commented on scripts/meta_analysis.py:
```diff
@@ -89,11 +113,18 @@ def inv_var_meta( studies : List[Tuple['Study','VariantData']], is_het_test = Fa
         het_p=None
-    return (beta_meta, math.sqrt(1/sum_inv_var), max(sys.float_info.min * sys.float_info.epsilon, 2 * scipy.stats.norm.sf(abs(sum(effs_inv_var) / math.sqrt(sum_inv_var) ))), het_p) if len(effs_inv_var)==len(studies) else None
-
+    return (beta_meta, math.sqrt(1/sum_inv_var), max(sys.float_info.min * sys.float_info.epsilon, 2 * scipy.stats.norm.sf(abs(sum(effs_inv_var) / math.sqrt(sum_inv_var) ))), het_p) if len(effs_inv_var)==len(studies) else (None, None, None, None)
```
the reason I don't like repeating theNone values to form a tuple is that a single None indicates that the data could not be computed at all. Returning None or possibly raising exceptions. In this way the caller can act accordingly.... it just needs to be documented that in case value can't be computed a None is returned.
Off course you could argue that you can check None values form the tuple... which of the values should you check? can some of them be None and some not? It is much more clear to return just None, so I would revert back to your previous implementation of checking if the value was none and then doing something on top of that instead of printing whatever the method returns in case when it can't compute the value.

juhaa replied:
Good point, I totally agree! I'll revert back to the original.

Fedja commented on scripts/meta_analysis.py:
```diff
@@ -59,11 +76,18 @@ def n_meta( studies : List[Tuple['Study','VariantData']], is_het_test = False ):
+            is_het_test: boolean, do heterogenity test
+        output:
+            tuple with results from meta-analysis and heterogeneity test p-value (if calculated)
+    '''
```
can you actually move the het test out of this function.  Each function should do only one thing. You can add the het test e.g. into  do_meta

juhaa replied:
Sure, sounds like a good improvement.

juhaa commented:
Should be fine now. Check if you agree with the changes I made in the do_meta to accompany the het_test function calculation there. Also the outputs from n, var and inv_var functions was changed a bit so we get all the needed variables for het test out. Output from do_meta stays the same though.

juhaa commented:
Upon further testing came across a new error which occurs when the left-most variant has zero or NA standard error:

```
Standard error was none/zero for variant chr:6 pos:31258364 ref:C alt:A beta:2.10489 pval:0.0 se:0.0  in study FINNGEN
Traceback (most recent call last):
  File "/META_ANALYSIS/scripts/meta_analysis.py", line 754, in <module>
    run()
  File "/META_ANALYSIS/scripts/meta_analysis.py", line 690, in run
    met = do_meta( matching_studies, methods=methods, is_het_test=args.is_het_test )
  File "/META_ANALYSIS/scripts/meta_analysis.py", line 509, in do_meta
    met = [ SUPPORTED_METHODS[m](study_list) for m in methods ]
  File "/META_ANALYSIS/scripts/meta_analysis.py", line 509, in <listcomp>
    met = [ SUPPORTED_METHODS[m](study_list) for m in methods ]
  File "/META_ANALYSIS/scripts/meta_analysis.py", line 106, in inv_var_meta
    beta_meta=sum(effs_inv_var)/ sum_inv_var
ZeroDivisionError: division by zero
```
Even though the SE is checked for 0 or NA, the fix is to break the loop, which in turn causes `sum_inv_var` to be zero and thus the ZeroDivisionError. See here:
https://github.com/juhaa/META_ANALYSIS/blob/e5f67cd40b73a74de6ab31cbe01f9da132689577/scripts/meta_analysis.py#L91-L106
I changed the break to return None instead since that would be ultimately returned anyways in case of zero or missing SE in any variant.

Fedja commented:
looks fine... now that we have actually variable number tuple returned and so many elements, for readability maybe turning to namedtuple or dataclass (https://docs.python.org/3/library/dataclasses.html) would be good.
